src/core/umount.c

```diff
@@ -379,6 +380,42 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+static int remount_with_timeout(MountPoint *m, usec_t timeout, char *options, int *n_failed, bool log_error) {
+        pid_t remount_pid;
+
+        int r;
+
+        /* Due to the possiblity of a remount operation hanging, we
+         * fork a child process and set a timeout. If the timeout
+         * lapses, the assumption is that that particular remount
+         * failed. */
+
+        remount_pid = fork();
+        if (remount_pid < 0)
+                return log_error_errno(errno, "Failed to fork: %m");
+
+        if (remount_pid == 0) {
+                /* Arm the alarm so that we can exit if the subsequent mount fails */
+                if (timeout != USEC_INFINITY)
+                        alarm((timeout + USEC_PER_SEC - 1) / USEC_PER_SEC);
```

Comment:
So, I am very sure alarm() is never the way to go for this. fs operations that hang are more often than not hanging in non-interruptable sleeps (NFS can deviate from that through "intr", but I am interested in the general case here). Hence I don't think relying on alarm()/SIGALRM is really the best bet here. I'd instead suggest building a careful loop around sigtimedwait()/waitid() in the parent, so that the parent decides how long to wait, not the child. And then, when the timeout is reached the child should be killed with SIGKILL, and then ignored if it still sticks around then.

Yupp, UNIX is messy, but let's do this right here, if we do it at all.